File: src/Y2018/facebook/Q071_Simplify_Path.java
package Y2018.facebook;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;

public class Q071_Simplify_Path {

    public String simplifyPath(String path) {
        Stack<String> stack = new Stack<>();
        Set<String> skip = new HashSet<>(Arrays.asList("..", ".", ""));
        for (String dir : path.split("/")) {
            if (dir.equals("..") && !stack.isEmpty()) {
                stack.pop();
            } else if (!skip.contains(dir)) {
                stack.push(dir);
            }
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.insert(0, "/" + stack.pop());
        }
        return sb.length() == 0 ? "/" : sb.toString();
    }

}

File: leetcode-go/Q071_Simplify_Path/Q071_Simplify_Path.go
package Q071_Simplify_Path

import (
	"strings"
)

func simplifyPath(path string) string {
	stack := []string{}
	for _, dir := range strings.Split(path, "/") {
		if dir == ".." && len(stack) != 0 {
			stack = stack[0 : len(stack)-1]
		}
		if dir != ".." && dir != "." && dir != "" {
			stack = append(stack, dir)
		}
	}
	if len(stack) == 0 {
		return "/"
	}
	return "/" + strings.Join(stack, "/")
}
